anatomy-of-locks-reduce.mdx

```diff
@@ -0,0 +1,128 @@
+
+Lock contention occurs when multiple transactions compete for access to the same database resource (like a table or row), and at least one transaction has to wait because it requires a lock that conflicts with locks held by other transactions. In PostgreSQL, this commonly happens during schema modifications (DDL operations) that require exclusive locks, or during heavy concurrent DML operations on the same rows. When contention occurs, transactions form a queue, waiting for their turn to acquire the needed lock. High lock contention can lead to decreased throughput, increased latency, and in severe cases, application timeouts or downtime. This is particularly problematic in high-traffic production systems where long-running queries or poorly timed schema changes can cause a cascade of waiting transactions, effectively blocking access to critical tables.
+
+## Reducing locking impact
+
+When making schema changes or running maintenance tasks in PostgreSQL, it's crucial to minimize locking to avoid blocking other queries and degrading performance. Not all operations require the same level of locking, and PostgreSQL offers tools and techniques to minimize locking impact.
+
+### Use CONCURRENTLY commands
+
+Commands like `CREATE INDEX CONCURRENTLY` or `ALTER TABLE DETACH PARTITION CONCURRENTLY` acquire less-restrictive locks compared to the same statements without `CONCURRENTLY`, allowing other operations to proceed. However, these commands:
+
+- Take longer to complete.
+- Are non-transactional (can't be in transaction block, can’t be rolled back).
+- Require additional care to handle failures, which can leave partial changes (there are commands like `FINALIZE` to clean up or finish the work).
+
+### Split complex operations
+
+One of the most effective strategies for reducing lock contention is breaking down large DDL operations into smaller, less-blocking steps.
+
+Let's imagine, you need to add a `NOT NULL` column with a default value:
+
+```sql
+ALTER TABLE mytable ADD COLUMN newcol timestamptz NOT NULL DEFAULT now();
+```
+
+This single command requires an **ACCESS EXCLUSIVE** lock and will rewrite the entire table. For large tables, this can lead to significant downtime as it:
+
+- **Blocks all concurrent access** (even `SELECT`s)
+- Holds the lock for **the entire duration of the table rewrite**
+- Can take minutes or hours for large tables
+
+Instead of a single heavy operation we chose above, we can break it into three less-blocking steps:
+
+```sql
+ALTER TABLE mytable ADD COLUMN newcol timestamptz DEFAULT now();
+UPDATE mytable SET newcol = now() WHERE newcol IS NULL;
+ALTER TABLE mytable ALTER COLUMN newcol SET NOT NULL;
+```
+
+1. First, add the nullable column with a default: `ALTER TABLE mytable ADD COLUMN newcol timestamptz DEFAULT now();`
+2. Then, populate any `NULL` values: `UPDATE mytable SET newcol = now() WHERE newcol IS NULL;` You should actually do this update in batches, remember any long-running query can cause problems.
```

Comment:
> You should actually do this update in batches, remember any long-running query can cause problems.

Do you think it's worth expanding on this, perhaps in a separate section?

`pgroll` does it's backfill in batches to avoid taking a row lock on every row in the table for example. I think it might be important enough to call out more than this one sentence.